### PR TITLE
Allow for other PythonScriptEngine environments besides task data

### PR DESCRIPTION
**Pull request by jbirddog**
~Opening this as a draft while I test more and work on back end integration to start to gather feedback on the approach. Names and new public apis might be tweaked along the way.~ Edit: ready for review.

The current usage of task data to track the environment for the PythonScriptEngine is convenient but does lend itself to growing task data as the environment is copied from task to task. Larger workflows can run into issues with their task data size. This change provides a way for applications utilizing SpiffWorkflow to provide their own mechanism for tracking the state of the PythonScriptEngine between executions. If nothing is done the library works as it currently does.

Box was moved from the PythonScriptEngine proper to the PythonScriptEngineEnvrionment since custom environments may or may not want to Box.

If a custom environment is provided the application needs to properly restore it like it would for a PythonScriptEngine instance.

### CODE REVIEW

File: SpiffWorkflow/bpmn/PythonScriptEngine.py [hunk 1]
```diff
@@ -26,66 +27,6 @@
 # 02110-1301  USA
 
 
-class Box(dict):
```
Comment (danfunk) [inline SpiffWorkflow/bpmn/PythonScriptEngine.py:29]: I like giving people the option of not using Box.

File: SpiffWorkflow/bpmn/PythonScriptEngine.py [hunk 2]
```diff
@@ -97,10 +38,14 @@ class PythonScriptEngine(object):
     expressions in a different way.
     """
 
-    def __init__(self, default_globals=None, scripting_additions=None):
-
-        self.globals = default_globals or {}
-        self.globals.update(scripting_additions or {})
+    def __init__(self, default_globals=None, scripting_additions=None, environment=None):
```
Comment (danfunk) [inline SpiffWorkflow/bpmn/PythonScriptEngine.py:42]: If the environment is passed in, why would continue to accept default_globals and default_additions as separate arguments?

Comment (danfunk) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:42]: for backwards compatibility of course.

Comment (danfunk) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:42]: Maybe this should log some sort of deprecation warning, so we can move people away from this in the future?

File: SpiffWorkflow/bpmn/PythonScriptEngineEnvironment.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import copy
+
+class BasePythonScriptEngineEnvironment:
```
Comment (danfunk) [inline SpiffWorkflow/bpmn/PythonScriptEngineEnvironment.py:4]: This is the interface I always wanted for SpiffWorkflows Python execution. Really beautiful.

**Review by danfunk — COMMENTED** [review]
I love this.

File: SpiffWorkflow/bpmn/PythonScriptEngine.py (hunk 2 above)
Comment (jbirddog) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:42]: Good call, will do.

File: SpiffWorkflow/bpmn/PythonScriptEngine.py (hunk 1 above)
Comment (essweine) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:29]: I think we should deprecate box too.  The ultimate goal here should be to provide a default that does nothing but execute and evaluate with the task data added to the context, with no additional changes to anything outside of execution/evaluation.

File: SpiffWorkflow/bpmn/PythonScriptEngine.py (hunk 2 above)
Comment (jbirddog) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:42]: Added the warning and updated tests that used default_globals/scripting_additions.

File: SpiffWorkflow/bpmn/PythonScriptEngine.py (hunk 1 above)
Comment (jbirddog) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:29]: Sounds good.

Comment (jbirddog) [reply SpiffWorkflow/bpmn/PythonScriptEngine.py:29]: Added a deprecation warning and a `BoxedTaskDataEnvironment` if consumers need/want it. Updated two tests that relied on Box to use this new environment.

**Review by essweine — APPROVED** [review]

**Pull request merged**